libflux/src/ast/walk/mod.rs

```diff
@@ -0,0 +1,363 @@
+            Expression::Binary(e) => Node::BinaryExpr(e),
+            Expression::Unary(e) => Node::UnaryExpr(e),
+            Expression::PipeExpr(e) => Node::PipeExpr(e),
+            Expression::Call(e) => Node::CallExpr(e),
+            Expression::Conditional(e) => Node::ConditionalExpr(e),
+            Expression::StringExpr(e) => Node::StringExpr(e),
+            Expression::Paren(e) => Node::ParenExpr(e),
+            Expression::Integer(e) => Node::IntegerLit(e),
+            Expression::Float(e) => Node::FloatLit(e),
+            Expression::StringLit(e) => Node::StringLit(e),
+            Expression::Duration(e) => Node::DurationLit(e),
+            Expression::Uint(e) => Node::UintLit(e),
+            Expression::Boolean(e) => Node::BooleanLit(e),
+            Expression::DateTime(e) => Node::DateTimeLit(e),
+            Expression::Regexp(e) => Node::RegexpLit(e),
+            Expression::PipeLit(e) => Node::PipeLit(e),
+            Expression::Bad(e) => Node::BadExpr(e),
+        }
+    }
+    fn from_stmt(stmt: &Statement) -> Node {
+        match stmt {
+            Statement::Expr(s) => Node::ExprStmt(s),
+            Statement::Variable(s) => Node::VariableAssgn(s),
+            Statement::Option(s) => Node::OptionStmt(s),
+            Statement::Return(s) => Node::ReturnStmt(s),
+            Statement::Bad(s) => Node::BadStmt(s),
+            Statement::Test(s) => Node::TestStmt(s),
+            Statement::Builtin(s) => Node::BuiltinStmt(s),
+        }
+    }
+    fn from_func_body(fb: &FunctionBody) -> Node {
+        match fb {
+            FunctionBody::Block(b) => Node::Block(b),
+            FunctionBody::Expr(e) => Node::from_expr(e),
+        }
+    }
+    fn from_prop_key(pk: &PropertyKey) -> Node {
+        match pk {
+            PropertyKey::Identifier(i) => Node::Identifier(i),
+            PropertyKey::StringLit(s) => Node::StringLit(s),
+        }
+    }
+    fn from_string_part(sp: &StringExprPart) -> Node {
+        match sp {
+            StringExprPart::Text(t) => Node::TextPart(t),
+            StringExprPart::Interpolated(e) => Node::InterpolatedPart(e),
+        }
+    }
+    fn from_assignment(a: &Assignment) -> Node {
+        match a {
+            Assignment::Variable(v) => Node::VariableAssgn(v),
+            Assignment::Member(m) => Node::MemberAssgn(m),
+        }
+    }
+}
+
+// Visitor defines a visitor pattern for walking the AST
+//
+// When used with the walk function, Visit will be called for every node
+// in depth-first order. After all children for a Node have been visted,
+// Done is called on that Node to signal that we are done with that Node.
+//
+// If Visit returns None, walk will not recurse on the children.
+pub trait Visitor: Sized {
+    // Visit is called for a node.
+    // The returned visitor will be used to walk children of the node.
+    // If visit returns None, walk will not recurse on the children.
+    fn visit(&self, node: &Node) -> Option<Self>;
```

Comment:
To clarify, the reason why this works is because if you attempt to do this with a unique pointer, it won't work correctly. So you instantiate the visitor trait using an `Rc` version of the type instead?

Reply:
Yes, after thinking it through the visitor trait itself doesn't care how you do it but you need to return an owned visitor from the visit function. Therefore if you want to reuse yourself you need multiple ownership  hence using `Rc` reference counting. 

I also ended up using RefCell to hide the mutability of the closure since its not part of the visitor itself.